### PR TITLE
refactor: 1348 upgrade to spring boot 3x

### LINKED ISSUE #1348: Upgrade to spring-boot 3.x
Upgrade from spring-boot 2.7.x to the latest spring-boot. 2.7 only receives minor fixes and patches. Some vulnerabilities are only fixed in 3.x. 

Additionally, support for spring-boot 2.7 runs out on the 2023-11-18.

Requires #1347 first.

### PR DESCRIPTION
**Pull request by TheGreatRefrigerator**
### Pull Request Checklist

- [x] 1. I have [**rebased**][rebase] the latest version of the master branch into my feature branch and all conflicts
         have been resolved.
- [x] 2. I have added information about the change/addition to functionality to the CHANGELOG.md file under the
         [Unreleased] heading.
- [x] 3. I have documented my code using JDocs tags.
- [x] 4. I have removed unnecessary commented out code, imports and System.out.println statements.
- [x] 5. I have written JUnit tests for any new methods/classes and ensured that they pass.
- [x] 6. I have created API tests for any new functionality exposed to the API.
- [ ] 7. If changes/additions are made to the ors-config.json file, I have added these to the [ors config documentation][config]
         along with a short description of what it is for, and documented this in the Pull Request (below).
- [ ] 8. I have built graphs with my code of the Heidelberg.osm.gz file and run the api-tests with all test passing
- [x] 9. I have referenced the Issue Number in the Pull Request (if the changes were from an issue).
- [ ] 10. For new features or changes involving building of graphs, I have tested on a larger dataset
          (at least Germany), and the graphs build without problems (i.e. no out-of-memory errors).
- [ ] 11. For new features or changes involving the graphbuilding process (i.e. changing encoders, updating the
          importer etc.), I have generated longer distance routes for the affected profiles with different options
          (avoid features, max weight etc.) and compared these with the routes of the same parameters and start/end
          points generated from the current live ORS.
          If there are differences then the reasoning for these **MUST** be documented in the pull request.
- [x] 12. I have written in the Pull Request information about the changes made including their intended usage
          and why the change was needed.
- [x] 13. For changes touching the API documentation, I have tested that the API playground [renders correctly][api].

Fixes # .

### Information about the changes
- Key functionality added: migration to spring-boot 3.1.1
- Reason for change: being up to date
As there were quite some refactorings done, this is not just a chore but refactor PR

### Examples and reasons for differences between live ORS routes, and those generated from this pull request
-

### Required changes to ors config (if applicable)
-

[config]: https://GIScience.github.io/openrouteservice/installation/Configuration.html
[api]: https://gitlab.gistools.geog.uni-heidelberg.de/giscience/openrouteservice-infrastructure/ors-docs-api#test-new-ors-documentation
[rebase]: https://github.com/GIScience/openrouteservice/blob/master/CONTRIBUTE.md#pull-request-guidelines

### CODE REVIEW

**Comment by TheGreatRefrigerator** [comment]
rebased on master

**Comment by MichaelsJP** [comment]
@TheGreatRefrigerator Looks like the docker setup is not working as before.
It changes from 000 (I guess no return) to 404. I guess the paths are somehow a bit different now when running the war deployment.

**Comment by TheGreatRefrigerator** [comment]
> @TheGreatRefrigerator Looks like the docker setup is not working as before. It changes from 000 (I guess no return) to 404. I guess the paths are somehow a bit different now when running the war deployment.

yes checking rn

**Comment by TheGreatRefrigerator** [comment]
@MichaelsJP docker workflow fixed by @koebi 's commit. Ready for review

**Comment by MichaelsJP** [comment]
This needs to wait until #1499 is merged. 

Can already be reviewed, nevertheless.

File: ors-api/src/main/resources/application.yml
Comment (MichaelsJP) [inline ors-api/src/main/resources/application.yml:167]: Why not 3.1 (for documentation purposes)?

Comment (TheGreatRefrigerator) [reply ors-api/src/main/resources/application.yml:167]: Because with version 3.1 the `type: object` is not generated for parameters. So for now we should stay at Open API 3.0

File: ors-engine/src/main/java/org/heigit/ors/jts/JTS.java
Comment (MichaelsJP) [inline ors-engine/src/main/java/org/heigit/ors/jts/JTS.java]:
```suggestion
        if (g instanceof LineString lineString) {
            return removeCollinearVertices(lineString);
```

File: ors-engine/src/main/java/org/heigit/ors/jts/JTS.java
Comment (MichaelsJP) [inline ors-engine/src/main/java/org/heigit/ors/jts/JTS.java]:
```suggestion
        if (geom instanceof LinearRing linearRing) {
```

File: ors-engine/src/main/java/org/heigit/ors/jts/JTS.java
Comment (MichaelsJP) [inline ors-engine/src/main/java/org/heigit/ors/jts/JTS.java]:
```suggestion
            Polygon poly = factory.createPolygon(linearRing, null);
```

File: ors-engine/src/main/java/org/heigit/ors/jts/JTS.java
Comment (MichaelsJP) [inline ors-engine/src/main/java/org/heigit/ors/jts/JTS.java]:
```suggestion
        if (geometry instanceof LineString lineString) {
            return removeCollinearVertices(lineString);
```

File: ors-engine/src/main/java/org/heigit/ors/matrix/algorithms/MatrixAlgorithmFactory.java
Comment (MichaelsJP) [inline ors-engine/src/main/java/org/heigit/ors/matrix/algorithms/MatrixAlgorithmFactory.java]:
```suggestion
        if (gh instanceof ORSGraphHopper orsGraphHopper) {
            if (req.getSearchParameters().getDynamicSpeeds() && orsGraphHopper.isCoreEnabled()) {
```

File: ors-engine/src/main/java/org/heigit/ors/routing/graphhopper/extensions/core/PrepareCore.java
Comment (MichaelsJP) [inline ors-engine/src/main/java/org/heigit/ors/routing/graphhopper/extensions/core/PrepareCore.java]:
```suggestion
        if (neighbors instanceof IntArrayList intArrayList)
            intArrayList.removeAll(isCoreNode);
```

File: ors-engine/src/main/java/org/heigit/ors/routing/graphhopper/extensions/edgefilters/core/LMEdgeFilterSequence.java
Comment (MichaelsJP) [inline ors-engine/src/main/java/org/heigit/ors/routing/graphhopper/extensions/edgefilters/core/LMEdgeFilterSequence.java]:
```suggestion
			if (edgeFilter instanceof AvoidBordersCoreEdgeFilter coreEdgeFilter){
```

File: ors-engine/src/main/java/org/heigit/ors/routing/graphhopper/extensions/edgefilters/core/LMEdgeFilterSequence.java
Comment (MichaelsJP) [inline ors-engine/src/main/java/org/heigit/ors/routing/graphhopper/extensions/edgefilters/core/LMEdgeFilterSequence.java]:
```suggestion
				return queryCountries.containsAll(Arrays.stream(coreEdgeFilter.getAvoidCountries()).boxed().collect(Collectors.toList()));
```

**Review by MichaelsJP — CHANGES_REQUESTED** [review]

**Comment by MichaelsJP** [comment]
Just a few minor Style things. Generally, it can become confusing when naming objects close to language level structurs such as `list`, `filter`, `string` etc.

**Comment by TheGreatRefrigerator** [comment]
changed some more of the pattern variables, see
https://github.com/GIScience/openrouteservice/compare/d38be18950e9f434f70cb7a4143ec1ffd16cd66d..61523f797ba36b216dc104f8f1e8a4b23816bd0d

**Review by MichaelsJP — APPROVED** [review]

**Review by MichaelsJP — APPROVED** [review]

**Pull request merged**
